### PR TITLE
Add H5Dchunk_iter()

### PR DESCRIPTION
**Pull request by ajelenak**
This new function is a faster way to go over all dataset's chunks and invoke a user callback function. It will also appear in the next HDF5-1.12.3 release, hence the mention of that version.

Not yet integrated in the high-level interface.

### CODE REVIEW

**Comment by mkitti** [comment]
I also have a patch to backport H5Dchunk_iter to the HDF5 1.10 branch:
https://github.com/HDFGroup/hdf5/pull/1968

I also have another patch to fix the doxygen example:
https://github.com/HDFGroup/hdf5/pull/2380

**Comment by ajelenak** [comment]
@mkitti I was told to reference 1.12.3 but will double-check.

**Comment by mkitti** [comment]
1.12.3 is correct:
https://github.com/HDFGroup/hdf5/pull/1970

**Comment by ajelenak** [comment]
I got a firm commitment that H5Dchunk_iter() will appear also in the next libhdf5-1.10.10 so I made the appropriate changes. Also added documentation for the new low-level API.

@tacaswell @takluyver I would like to merge this PR early next week now that libhdf5-1.14.0 is out.

**Comment by ajelenak** [comment]
In my own usage, and it's been just over a year since I brought that function in my custom h5py, the average speed up was about 2x compared to the functions you mention. I agree it is a bit more involving with a callback function but not that much since the same approach is used to iterate over groups and datasets.

**Comment by mkitti** [comment]
The problem with `get_chunk_info` is that on each call you iterate through the B-tree structure every single time to find the chunk you are interested in.

If you used `get_chunk_info` to retrieve chunk information for every chunk using a for loop, you would iterate over `sum(range(N+1))`, or  `N*(N+1)/2`, leaves which is $O(N^2)$.

With `chunk_iter` you just have to iterate over N items to iterate over all chunks.

See https://github.com/HDFGroup/hdf5/pull/6 for the original explanation.

**Comment by jreadey** [comment]
@ajelenak - would it be possible to have a fixed callback implemented in cython that fills in an array of chunk infos?  So the client code doesn't implement the callback but just gets a list of chunk locations.  Guess that would avoid the callback overhead that @takluyver pointed out.

File: h5py/h5d.pyx [hunk 1]
```diff
@@ -66,6 +66,83 @@ IF HDF5_VERSION >= (1, 10, 5):
...
+            chunk_info = StoreInfo(None, filter_mask, None, size)
+
+        visit.retval = visit.func(chunk_info)
+        if visit.retval is not None:
+            return 1
+        return 0
+
+
+    @with_phil
+    def visitchunks(DatasetID dsid not None, object func, PropID dxpl=None):
```
[the remark this answers is not in the document]
Comment (ajelenak) [reply h5py/h5d.pyx]: You are correct, I made this change.

**Comment by ajelenak** [comment]
I think it should be doable but let's first get this branch merged and then we can think of something like what you suggest.

**Comment by mkitti** [comment]
Use of `H5Dchunk_iter` was implemented in https://github.com/h5py/h5py/pull/2202 as a low level function `h5py.h5d.DatasetID.chunk_iter`

**Pull request closed without merging**